src/services/DomainModel.ts

```diff
@@ -0,0 +1,125 @@
+// Do not assume you can reach every node, sensor, or reading starting from a
+// gateway you're looking at. Be mindful that a QueryFilter is probably in
+// effect which has limited the extent to which we have populated all the
+// objects into memory. For example, if you're working with recent data from the
+// 'Black Pearl' node you might have info about the Gateway, but just the one
+// node, and a few days of readings. You'll find `Unknown`s at the points in the
+// tree that we don't care about in the context of an active Query.
+
+export const unknown = Symbol("Unknown");
+export type Unknown = typeof unknown;
+
+// These `...ID` types prevent `let gatewayID = getNodeID()` which would sadly
+// compile if each were simply string. Is there an easier way? Is it worth it?
+export type GatewayID = { type: "GatewayID"; id: string };
+export type NodeID = { type: "NodeID"; id: string };
+export type ReadingSchemaID = { type: "ReadingSchemaID"; id: string };
+
+// TODO (carl): So the types don't need to be peppered with `| Unknown`:
+//
+// - we might be able to make a type like Shallow<Node> that's not guaranteed to
+//   link to every other object. I doubt it would be worth the complication.
+//
+// - we could just use `?` and `| undefined` and explain that undefined means
+//   something like "idk" or "Unknown". I just hate comments explaining
+//   things when good naming will make semantics obvious. Is it a common enough
+//   practice in JS? For example `sensors?: Set<Sensor | undefined>;`
+
+export interface Gateway {
+  // Attributes
+  name: string;
+  description: string;
+  shortDescription: string;
+  // Links
+  nodes: Set<Node> | Unknown;
+  // Implementation Details
+  readonly id: GatewayID; // Reference across the whole system architecture
+  type: "Gateway"; // For typescript narrowing and javascript debugging
+}
+
+export interface Node {
+  // Attributes
+  description: string;
+  shortDescription: string;
+  locationHref: string;
+  locationName: string;
+  name: string;
+  // Links
+  gateway: Gateway;
+  sensors: Set<Sensor> | Unknown;
+  // Implementation Details
+  readonly id: NodeID;
+  type: "Node";
+}
+
+export interface Sensor {
+  // Attributes
+  // Links
+  node: Node;
+  readings: Set<Reading> | Unknown;
+  schema: ReadingSchema;
+  // Implementation Details
+  type: "Sensor";
+}
+
+export interface ReadingSchema {
+  // Attributes
+  measure: string; // Temperature, etc.
+  name: string; // Outdoor Temperature, etc.
+  systemOfMeasurement: string; // SI, Imperial, US Customary, etc.
+  unit: string; // Kelvin, etc.
+  unitSymbol: string; // K, etc.
+  // Links
+  // Implementation Details
+  readonly id: ReadingSchemaID;
+  type: "ReadingSchema";
+}
+
+export interface Reading {
+  // Attributes
+  value: JSONValue;
```

Comment:
I don't mind this, but if the pain is too much, how about simply `unknown` and we use knowledge of the schema to know the correct type?